### PR TITLE
feat(api): extend controller, search criteria progress

### PR DESCRIPTION
**Pull request by oXtxNt9U**
## Checklist

- [ ] Documentation _(if necessary)_
- [ ] Tests _(if necessary)_
- [ ] Ready to be merged

### CODE REVIEW

File: packages/api-http/source/controllers/blocks.ts [hunk 1]
```diff
@@ -109,4 +125,9 @@ export class BlocksController extends Controller {
 
 	// 	return block;
 	// }
+
+	private getBlockCriteriaByIdOrHeight(idOrHeight: string): Search.Criteria.OrBlockCriteria {
+		const asHeight = Number(idOrHeight);
+		return asHeight && asHeight <= 50_000_000 ? { height: asHeight } : { id: idOrHeight };
```
Comment (sebastijankuzner) [inline packages/api-http/source/controllers/blocks.ts:131]: Is _asHeight <= 50_000_000_ check necessary?

File: packages/api-http/source/controllers/transactions.ts [hunk 2]
```diff
@@ -7,13 +7,13 @@ import { Controller } from "./controller";
...
-	@inject(ApiDatabaseIdentifiers.TransactionRepository)
-	private readonly transactionRepository!: ApiDatabaseContracts.ITransactionRepository;
+	@inject(ApiDatabaseIdentifiers.TransactionRepositoryFactory)
+	private readonly transactionRepositoryFactory!: ApiDatabaseContracts.ITransactionRepositoryFactory;
 
 	public async index(request: Hapi.Request, h: Hapi.ResponseToolkit) {
 		const pagination = this.getQueryPagination(request.query);
 
-		const [transactions, totalCount] = await this.transactionRepository
+		const [transactions, totalCount] = await this.transactionRepositoryFactory()
```
Comment (sebastijankuzner) [inline packages/api-http/source/controllers/transactions.ts:16]: Can we initialize repositories on controller creation?

File: packages/api-database/source/repositories/block-repository.ts [hunk 3]
```diff
@@ -1,9 +1,61 @@
...
 export const makeBlockRepository = (dataSource: RepositoryDataSource): IBlockRepository =>
-	dataSource.getRepository(Block).extend({
+	makeExtendedRepository<Block, IBlockRepositoryExtension>(Block, dataSource, {
+		async findManyByCriteriaJoinTransactions(
+			// @ts-ignore
+			transactionRepository: ITransactionRepository,
+			blockCriteria: Criteria.OrBlockCriteria,
+		): Promise<Block[]> {
+			const blockExpression = await BlockFilter.getExpression(blockCriteria);
+			//	const transactionBlockCriteria = blockModels.map((b) => ({ blockId: b.id }));
```
[the remark this answers is not in the document]
Comment (oXtxNt9U) [reply packages/api-database/source/repositories/block-repository.ts:20]: Yeah, just WIP will be taken care of in follow up PRs

File: packages/api-http/source/controllers/transactions.ts (hunk 2 above)
Comment (oXtxNt9U) [reply packages/api-http/source/controllers/transactions.ts:16]: As discussed, `QueryHelper` makes the repositories stateful. So a copy is needed per request (which should be cheap). Made a task to eventually improve it if needed.

**Review by sebastijankuzner — APPROVED** [review]

**Pull request merged**